vowpalwabbit/parser.h

```diff
@@ -35,6 +37,46 @@ struct example_initializer
   example* operator()(example* ex);
 };
 
+struct IO_State {
+
+      std::queue<IO_Item> *io_lines = nullptr;
+      std::atomic<bool> done_with_io;
+
+      IO_State(){
```

Comment:
One of these two constructors should be removed. Otherwise you'll end up with 2 different situations to clean up
1. we call `new` on io_lines. IO_State now owns the memory and needs to call `delete`.
2. the io_lines pointer is passed in. IO_State does not own the memory, and should not call `delete`.

This will cause either a memory leak or a double free exception.

If IO_State is being shared between the threads (which looks to be the case) then the first scenario is the one we want to use and the queue should never be passed in as a parameter or even be accessible outside of the class.